### PR TITLE
fix(core): removed xxhash to fix #11071

### LINKED ISSUE #11071: "perf(core): change sha1 to xxhash" commit broke runtime of Nest.js application on Ubuntu 20.04
### Did you read the migration guide?

- [X] I have read the whole migration guide

### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Potential Commit/PR that introduced the regression

7f43821d4810e6b2eff87f159eac2a8177f2736a

### NestJS version

9.3.2 -> 9.3.5

### Describe the regression

So, I've upgraded @nestjs/common and @nestjs/core dependencies from 9.3.2. to 9.3.5. Decided to run them on my PC with Windows 10, runs perfectly, all good. Pushed to my Github and let the Actions complete the build and deployment processes, and this is "docker logs" output of the container that runs my Nest.js application.

```
Error: Cannot find module '@node-rs/xxhash-linux-x64-gnu'
Require stack:
- /app/node_modules/@node-rs/xxhash/index.js
- /app/node_modules/@nestjs/core/inspector/deterministic-uuid-registry.js
- /app/node_modules/@nestjs/core/inspector/uuid-factory.js
- /app/node_modules/@nestjs/core/injector/instance-wrapper.js
- /app/node_modules/@nestjs/core/injector/injector.js
- /app/node_modules/@nestjs/core/injector/module-ref.js
- /app/node_modules/@nestjs/core/injector/lazy-module-loader/lazy-module-loader.js
- /app/node_modules/@nestjs/core/inspector/serialized-graph.js
- /app/node_modules/@nestjs/core/injector/container.js
- /app/node_modules/@nestjs/core/injector/index.js
- /app/node_modules/@nestjs/core/index.js
- /app/dist/main.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1047:15)
    at Module._load (node:internal/modules/cjs/loader:893:27)
    at Module.require (node:internal/modules/cjs/loader:1113:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/app/node_modules/@node-rs/xxhash/index.js:172:31)
    at Module._compile (node:internal/modules/cjs/loader:1226:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
    at Module.load (node:internal/modules/cjs/loader:1089:32)
    at Module._load (node:internal/modules/cjs/loader:930:12)
    at Module.require (node:internal/modules/cjs/loader:1113:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/app/node_modules/@node-rs/xxhash/index.js',
    '/app/node_modules/@nestjs/core/inspector/deterministic-uuid-registry.js',
    '/app/node_modules/@nestjs/core/inspector/uuid-factory.js',
    '/app/node_modules/@nestjs/core/injector/instance-wrapper.js',
    '/app/node_modules/@nestjs/core/injector/injector.js',
    '/app/node_modules/@nestjs/core/injector/module-ref.js',
    '/app/node_modules/@nestjs/core/injector/lazy-module-loader/lazy-module-loader.js',
    '/app/node_modules/@nestjs/core/inspector/serialized-graph.js',
    '/app/node_modules/@nestjs/core/injector/container.js',
    '/app/node_modules/@nestjs/core/injector/index.js',
    '/app/node_modules/@nestjs/core/index.js',
    '/app/dist/main.js'
  ]
}
```

### Minimum reproduction code

_No response_

### Input code

```ts

```

### Expected behavior

Application should run normally on Ubuntu 20.04, just like it does on Windows 10.

### Other

Runs on Github hosted Ubuntu runners, this is Actions response for a simple unit test stage:
image

Produces pretty much the same error:

```
Require stack:
      node_modules/@node-rs/xxhash/index.js
      node_modules/@nestjs/core/inspector/deterministic-uuid-registry.js
      node_modules/@nestjs/core/inspector/uuid-factory.js
      node_modules/@nestjs/core/injector/instance-wrapper.js
      node_modules/@nestjs/core/injector/injector.js
      node_modules/@nestjs/core/injector/module-ref.js
      node_modules/@nestjs/core/injector/lazy-module-loader/lazy-module-loader.js
      node_modules/@nestjs/core/inspector/serialized-graph.js
      node_modules/@nestjs/core/injector/container.js
      node_modules/@nestjs/testing/testing-module.builder.js
      node_modules/@nestjs/testing/test.js
      node_modules/@nestjs/testing/index.js
      src/tests/auth.controller.spec.ts

    > 1 | import { Test, TestingModule } from "@nestjs/testing";
        | ^
      2 | import { I18nContext } from "nestjs-i18n";
      3 | import { getMockReq, getMockRes } from "@jest-mock/express";
      4 | import { UserService } from "@user/services";

      at Resolver._throwModNotFoundError (node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.<anonymous> (node_modules/@node-rs/xxhash/index.js:[17](https://github.com/Pinehood/ph-s2u-backend/actions/runs/4116536300/jobs/7106703742#step:7:18)2:31)
      at Object.<anonymous> (node_modules/@nestjs/core/inspector/deterministic-uuid-registry.js:4:[18](https://github.com/Pinehood/ph-s2u-backend/actions/runs/4116536300/jobs/7106703742#step:7:19))
      at Object.<anonymous> (node_modules/@nestjs/core/inspector/uuid-factory.js:5:39)
      at Object.<anonymous> (node_modules/@nestjs/core/injector/instance-wrapper.js:11:24)
      at Object.<anonymous> (node_modules/@nestjs/core/injector/injector.js:14:28)
      at Object.<anonymous> (node_modules/@nestjs/core/injector/module-ref.js:8:[20](https://github.com/Pinehood/ph-s2u-backend/actions/runs/4116536300/jobs/7106703742#step:7:21))
      at Object.<anonymous> (node_modules/@nestjs/core/injector/lazy-module-loader/lazy-module-loader.js:5:22)
      at Object.<anonymous> (node_modules/@nestjs/core/inspector/serialized-graph.js:8:30)
      at Object.<anonymous> (node_modules/@nestjs/core/injector/container.js:6:28)
      at Object.<anonymous> (node_modules/@nestjs/testing/testing-module.builder.js:6:[21](https://github.com/Pinehood/ph-s2u-backend/actions/runs/4116536300/jobs/7106703742#step:7:22))
      at Object.<anonymous> (node_modules/@nestjs/testing/test.js:5:34)
      at Object.<anonymous> (node_modules/@nestjs/testing/index.js:11:[22](https://github.com/Pinehood/ph-s2u-backend/actions/runs/4116536300/jobs/7106703742#step:7:23))
      at Object.<anonymous> (src/tests/auth.controller.spec.ts:1:1)
```

### PR DESCRIPTION
**Pull request by H4ad**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #11071

## What is the new behavior?

Removed xxhash because they could break the CI system because the way of binary packages are bundled and installed by NPM/Yarn.

About the performance implications, using `sha1` could be a little bit slower but still faster than the original implementation with `uuid` for registry and `object-hash` for module token factory.

```
uuid v4 x 14,449,798 ops/sec ±2.08% (75 runs sampled)
uuid v5 x 182,749 ops/sec ±1.26% (80 runs sampled)
sha1 x 396,744 ops/sec ±3.76% (72 runs sampled)
node-rs/xxHash xxh32 x 3,104,990 ops/sec ±1.90% (76 runs sampled)
Fastest is uuid v4
```

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/injector/module-token-factory.ts [hunk 1]
```diff
@@ -64,7 +64,7 @@
   }
 
   private hashString(value: string): string {
-    return xxh32(value).toString();
+    return createHash('sha1').update(value).digest('hex');
```
[the remark this answers is not in the document]
Comment (kamilmysliwiec) [reply packages/core/injector/module-token-factory.ts:67]: Let's do this (since it's faster)

**Comment by kamilmysliwiec** [comment]
I'm wondering if we should a simple API to swap the underlying hashing algorithm with something else - if needed. In this case, some could intentionally switch to use xxhash if they're aware of what they're doing (npm i instead of CI in CI/CD etc etc). What do you think? For this we'd need a bit more realistic benchmarks though - compare real-world (or at least a sample but larger application) with sha vs xxhash - just to give folks an idea of what impact that could have. Do you have a random app locally that you could test that against?

**Comment by H4ad** [comment]
I think is a good idea, maybe we could have other packages that implement `ModuleTokenFactory` entirely.

Also, about the hash algorithms, I did some research in the past and I found this one: https://gist.github.com/hyamamoto/fd435505d29ebfa3d9716fd2be8d42f0

```
hashCode x 6,966,087 ops/sec ±1.42% (88 runs sampled)
sha1 x 554,580 ops/sec ±4.45% (77 runs sampled)
sha256 x 524,163 ops/sec ±4.55% (73 runs sampled)
sha512 x 435,163 ops/sec ±5.01% (75 runs sampled)
```

For `DeterministicUuidRegistry`, I think we could use this simple hash function because even with collision, we already solve that by checking if the hash already was generated (not in the new version). We only need to worry about how often the collision is for the values that the API could see.

If an API has too much collision with `hashCode`, we could have an implementation of `ModuleTokenFactory` that is in the middle of the performance, like `sha256`, and we could export (in other packages) other variants that have different performances, like `hashCode` and `xxhash`.

About the random app, I don't have but is easier to create one, `ModuleTokenFactory` it only receives two types of values: with or without metadata. We only need to mock simple values, without metadata, and pass a larger object with metadata.

**Comment by kamilmysliwiec** [comment]
> For DeterministicUuidRegistry, I think we could use this simple hash function because even with collision, we already solve that by checking if the hash already was generated (not in the new version). We only need to worry about how often the collision is for the values that the API could see.

Let's give it a shot! 

As for the API for swapping different hash algorithms, let's think about this once we have some real-world benchmarks to see if it even makes sense (how much of a difference that would be - time-wise)

**Comment by H4ad** [comment]
@kamilmysliwiec Okay, I will do some benchmarks and I push the changes with the new hash function.

Also, this line:

https://github.com/nestjs/nest/blob/bd92969b0c713cbfafdeee9b84e8503eb100471e/packages/core/injector/module-token-factory.ts#L19

The implementation is:

https://github.com/nestjs/nest/blob/bd92969b0c713cbfafdeee9b84e8503eb100471e/packages/core/injector/module-token-factory.ts#L53-L59

If your goal is to have deterministic ids, we should not change how we get the `moduleId`, to use `DeterministicUuidRegistry` too?

**Comment by H4ad** [comment]
About changing to `hashCode`, the profiler information of `9.3.5`:

```
Statistical profiling result from perf-startup-10000_baseline_9_3_5.log, (30428 ticks, 160 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
  24339   80.0%          /home/h4ad/.asdf/installs/nodejs/16.16.0/bin/node
    211    0.7%          /usr/lib/x86_64-linux-gnu/libc-2.31.so
      6    0.0%          /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
      1    0.0%          [vdso]

 [JavaScript]:
   ticks  total  nonlib   name
    420    1.4%    7.2%  LazyCompile: *InstanceWrapper /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/injector/instance-wrapper.js:17:16
    269    0.9%    4.6%  LazyCompile: *reflectKeyMetadata /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/scanner.js:164:23
    269    0.9%    4.6%  LazyCompile: *OrdinaryGetMetadata /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/reflect-metadata/Reflect.js:600:37
    168    0.6%    2.9%  LazyCompile: *reflectDynamicMetadata /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/scanner.js:117:27
    147    0.5%    2.5%  LazyCompile: *next /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/iterare/lib/filter.js:8:9
    145    0.5%    2.5%  LazyCompile: *getMetadata /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/reflect-metadata/Reflect.js:352:29
    141    0.5%    2.4%  LazyCompile: *resolveConstructorParams /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/injector/injector.js:96:35
    139    0.5%    2.4%  LazyCompile: *decirc /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/fast-safe-stringify/index.js:63:17
    123    0.4%    2.1%  LazyCompile: *loadInstance /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/injector/injector.js:31:23
```

And the new profiler:

```
Statistical profiling result from perf-startup-10000_baseline_9_3_5_hashCode.log, (30948 ticks, 136 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
  24941   80.6%          /home/h4ad/.asdf/installs/nodejs/16.16.0/bin/node
    243    0.8%          /usr/lib/x86_64-linux-gnu/libc-2.31.so
      9    0.0%          /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
      5    0.0%          [vdso]

 [JavaScript]:
   ticks  total  nonlib   name
    378    1.2%    6.6%  LazyCompile: *InstanceWrapper /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/injector/instance-wrapper.js:17:16
    367    1.2%    6.4%  LazyCompile: *getMetadata /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/reflect-metadata/Reflect.js:352:29
    288    0.9%    5.0%  LazyCompile: *OrdinaryGetMetadata /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/reflect-metadata/Reflect.js:600:37
    179    0.6%    3.1%  LazyCompile: *reflectDynamicMetadata /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/scanner.js:117:27
    165    0.5%    2.9%  LazyCompile: *reflectInjectables /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/scanner.js:134:23
    143    0.5%    2.5%  LazyCompile: *loadInstance /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/injector/injector.js:31:23
    129    0.4%    2.2%  LazyCompile: *decirc /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/fast-safe-stringify/index.js:63:17
    127    0.4%    2.2%  LazyCompile: *resolveConstructorParams /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/injector/injector.js:96:35
    121    0.4%    2.1%  LazyCompile: *scanForModules /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/node_modules/@nestjs/core/scanner.js:33:25
```

The version with `hashCode` is a little bit slower (in terms of ticks), but it helps to reduce the ticks from `InstanceWrapper`.
In terms of initialization, it went from `3.217948251900077ms` to `3.2779356801003217ms`.

To compare, using `crypto.createHash('sha256')`:

```


And the time spent is: 3.2936103528998792ms

**Comment by H4ad** [comment]
I did some benchmarks with `hashCode`:

Time Spent to initialize (10k times):
- hashCode: 20.566240562300383ms, total time: 0h:03m:33s
- xxhash: 11.35147113860026ms, total time: 0h:01m:59s
- sha256: 12.276356884299963ms, total time: 0h:02m:09s

> I create a very large module, like, 64kb * 20 modules, and based on the results below, that's why it was so slow.

Based on that I change the benchmarks to have string length levels, here is the result:

```
small (41): hashCode x 6,925,308 ops/sec ±0.96% (88 runs sampled)
small (41): sha256 x 584,153 ops/sec ±1.50% (92 runs sampled)
small (41): node-rs/xxHash xxh32 x 3,614,564 ops/sec ±0.58% (92 runs sampled)

medium (123): hashCode x 2,354,178 ops/sec ±4.32% (88 runs sampled)
medium (123): sha256 x 530,818 ops/sec ±2.16% (90 runs sampled)
medium (123): node-rs/xxHash xxh32 x 2,882,828 ops/sec ±0.53% (88 runs sampled)

large (246): hashCode x 1,215,748 ops/sec ±0.90% (87 runs sampled)
large (246): sha256 x 531,031 ops/sec ±0.97% (90 runs sampled)
large (246): node-rs/xxHash xxh32 x 2,711,832 ops/sec ±0.62% (91 runs sampled)

ridiculous (492): hashCode x 588,048 ops/sec ±0.95% (90 runs sampled)
ridiculous (492): sha256 x 411,592 ops/sec ±3.38% (84 runs sampled)
ridiculous (492): node-rs/xxHash xxh32 x 2,062,809 ops/sec ±0.80% (86 runs sampled)

ridiculous1 (615): hashCode x 470,271 ops/sec ±0.56% (91 runs sampled)
ridiculous1 (615): sha256 x 394,127 ops/sec ±3.48% (82 runs sampled)
ridiculous1 (615): node-rs/xxHash xxh32 x 1,787,767 ops/sec ±0.49% (90 runs sampled)

ridiculous2 (820): hashCode x 357,268 ops/sec ±0.59% (89 runs sampled)
ridiculous2 (820): sha256 x 384,852 ops/sec ±3.81% (83 runs sampled)
ridiculous2 (820): node-rs/xxHash xxh32 x 1,552,160 ops/sec ±0.74% (92 runs sampled)

ridiculous3 (1230): hashCode x 241,444 ops/sec ±0.48% (90 runs sampled)
ridiculous3 (1230): sha256 x 380,965 ops/sec ±0.74% (89 runs sampled)
ridiculous3 (1230): node-rs/xxHash xxh32 x 1,218,371 ops/sec ±0.85% (89 runs sampled)
Fastest is small (41): hashCode
```

So, the speed will depend on the size of the string, the only constant is: `xxhash` is faster in all cases (except from small)

I put this `console.log(JSON.stringify(opaqueToken).length);` to check the size of the tokens and what I got is:

```
6260
79
810
765
765
696
79
79
90
96
82
84
80
85
289
79
249
79
95
80
80
912
88
80
87
88
81
100
93
1024
87
99
92
88
976
89
82
932
86
79
904
94
87
970
93
86
962
91
84
946
94
87
976
81
920
91
84
948
99
92
1014
98
91
1006
96
89
990
109
102
86
964
94
1030
96
1326
80
1010
1010
101
93
88
88
81
920
89
87
86
91
99
94
92
97
89
96
92
84
91
85
85
80
912
93
100
93
84
944
89
988
89
988
96
96
97
100
91
95
87
810
696
79
82
80
85
95
87
88
100
87
99
89
86
94
93
91
94
91
99
98
96
... and so on
```

> All values: [size-of-strings.txt](https://github.com/nestjs/nest/files/10680066/size-of-strings.txt)

So, the main question is, worth to have this function over `sha256`? For most cases, yes, for some cases, not.
Maybe we could put a stage hash, so we less than `450`, we use `hashCode`, and for the other values we use `sha256`.

Tell me what you think. Anyway, I think these optimizations should be in another PR, this one we can close in favor to fix the problem with users CI.

**Comment by kamilmysliwiec** [comment]
> Tell me what you think. Anyway, I think these optimizations should be in another PR, this one we can close in favor to fix the problem with users CI.

Agree. Let's think about this in the future & in the following PRs.

**Comment by Dzixxx** [comment]
@kamilmysliwiec just as an FYI this also broke build process of standalone app of nest-commander (@jmcdo29) - yes I was creating one big js file to not install any dependencies to boost github actions execution (from 2mins to 8seconds just to reduce billing costs) 😂

**Pull request merged**